spring-cloud-kubernetes-commons/src/main/java/org/springframework/cloud/kubernetes/commons/config/SecretsConfigProperties.java

```diff
@@ -158,13 +162,30 @@ public boolean isEmpty() {
 			return !StringUtils.hasLength(this.name) && !StringUtils.hasLength(this.namespace);
 		}
 
-		public SecretsConfigProperties.NormalizedSource normalize(String defaultName, String defaultNamespace,
-				Map<String, String> defaultLabels) {
+		public Stream<NormalizedSource> normalize(String defaultName, String defaultNamespace,
+				Map<String, String> defaultLabels, Environment environment) {
+
+			Stream.Builder<NormalizedSource> normalizedSources = Stream.builder();
+
 			String normalizedName = StringUtils.hasLength(this.name) ? this.name : defaultName;
 			String normalizedNamespace = StringUtils.hasLength(this.namespace) ? this.namespace : defaultNamespace;
 			Map<String, String> normalizedLabels = this.labels.isEmpty() ? defaultLabels : this.labels;
 
-			return new SecretsConfigProperties.NormalizedSource(normalizedName, normalizedNamespace, normalizedLabels);
+			// if users do not specify a name for the secret (normalizedName), we still
```

Comment:
this is the crox of the change/fix. 

What we used to have:

```
spring:
  cloud:
    kubernetes:
      secrets:
         namespace: spring-k8s
         name: some-secret
         labels:
            a: b 
```

With a configuration as above, we used to read : 

- a secret with name `some-secret` 
- all secrets with lables `a=b`. Notice that in this case, we did not care about the name of the secret. 

Same happens after this change, but this time, we have separated such a configuration in _two_ normalized sources (as opposed to only one before). This allows us to : 

- fix the actual issue of reading the a secret only once (you can look at the bug description for more details + I will add a few unit test to enforce this soon). This holds because we make a distinction now for sources that are "name based" and "labels based"; and if we do not care about `lables` in the case of "name based secrets" (and in reverse), we can always do:
 
```
// named-base source has Collections.emptyMap(), which denotes "we do not care about lables"
new NormalizedSource(secretName, normalizedNamespace, Collections.emptyMap());
```

and 

```
// lables-base source has null first argument, which denotes "we do not care about secret name"
new NormalizedSource(null, normalizedNamespace, labels);
```

Reply:
Sorry for the confusion, I meant a constructor for NormalizedSource for the name case and the labels case.  I kind of dislike passing `null` and an empty `Map` to signify these different use cases.

Reply:
every time I want to fix these kind of issues, this problem pops up again and again for me. I really want to follow your advise, and the changes I just added confirm that. But, as you can see - this is involved. Do I want to keep them the way I have them now? Of course I do - clear separation, easy to add new features, maintainable, etc. Will you agree with such breaking changes? Well, I think we both know the answer :) At least for sure not in `2.0.x`. Be it my will, I would leave this defect open (its just an extra read of a secret, potentially for some users, not that of a big deal) and instead take this re-factoring in. It's not just about the fact that I wrote it, I don't really care about that; but it puts our code base in a much better shape and it fixes many issues/inconsistencies we currently have.

> I kind of dislike passing null and an empty Map to signify these different use cases.

YES! You make perfect sense, but as soon as you go into the details, things complicate badly, as this updated PR shows. 

This PR will of course fail, but I've added the changes (in fabric8 only), so that you could have an understanding where I would take this.

Reply:
I've pushed an implementation for fabric8 that works and passes all the tests, I will not do the same for k8s-client until I get some feedback. Otherwise I might invest a lot of time into something that will end up nowhere. thank you for looking into this!

Reply:
the idea is not complicated at all.

- create a `NormalizedSource`, from it create `NamedSecretNormalizedSource` and `LabeledSecretNormalizedSource`.
- define a type called `NormalizedSourceType`
- instead of passing lots of arguments around, create a `Fabric8ConfigContext`
- With the above create such a strategy pattern: 

```
EnumMap<NormalizedSourceType, Function<Fabric8ConfigContext, Map.Entry<String, Map<String, Object>>>> STRATEGIES = ...
```

apply each strategy type for each `NormalizedSourceType`. It's doing the same exact things it did (with some fixes that are not transparent), but in a way that makes the code easier to extend. 

If you ever agree that this can go in, in some release, same changes should be done for config maps, to stay on par - though it is not mandatory, as config maps do not have labels selector currently.